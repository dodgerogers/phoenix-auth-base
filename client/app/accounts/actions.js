import { actionTypes } from './constants';

export const getCurrentUserRequest = () => ({
  type: actionTypes.GET_CURRENT_USER_REQUEST,
  accessToken,
});

export const getCurrentUserSuccess = user => ({
  type: actionTypes.GET_CURRENT_USER_SUCCESS,
  user,
});

export const getCurrentUserFailure = () => ({
  type: actionTypes.GET_CURRENT_USER_FAILURE,
});

export const getCurrentUserProfilesRequest = () => ({
  type: actionTypes.GET_CURRENT_USER_PROFILES_REQUEST,
  accessToken,
});

export const getCurrentUserProfilesSuccess = profiles => ({
  type: actionTypes.GET_CURRENT_USER_PROFILES_SUCCESS,
  profiles,
});

export const getCurrentUserProfilesFailure = () => ({
  type: actionTypes.GET_CURRENT_USER_PROFILES_FAILURE,
});

export const setCurrentUserProfile = (profileId) => ({
  type: actionTypes.SET_CURRENT_USER_PROFILE,
  profileId,
})
