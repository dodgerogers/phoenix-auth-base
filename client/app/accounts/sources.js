import HTTP from '../lib/utils/HTTP';

export const currentUser = () => {
  return HTTP.get('/api/current_user');
};

export const currentUserProfiles = () => {
  return HTTP.get('/api/current_user/profiles');
};
