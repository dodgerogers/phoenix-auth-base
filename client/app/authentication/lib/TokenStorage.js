import Cookies from 'js-cookie';
import moment from 'moment';


export const COOKIE_KEY = 'token';
export const generateExpiry = token => moment(token.createdAt).add(token.expiresIn, 'seconds').toDate();
export const encode = token => btoa(JSON.stringify(token));
export const decode = token => {
  try {
    return JSON.parse(atob(token));
  } catch (err) {
    return null;
  }
};

export const store = token => {
  return new Promise((resolve, reject) => {
    const expires = generateExpiry(token);
    const cookie = Cookies.set(COOKIE_KEY, encode(token), { expires });

    if (cookie) {
      return resolve(cookie);
    }

    reject(null);
  });
}

export const fetch = () => {
  return new Promise((resolve, reject) => {
    const encodedToken = Cookies.get(COOKIE_KEY);

    if (encodedToken) {
      const decodedToken = decode(encodedToken);
      return resolve(decodedToken);
    }

    reject(null);
  });
};

export const remove = () => {
  return new Promise((resolve, reject) => {
    try {
      const result = Cookies.remove(COOKIE_KEY);
      resolve(result);
    } catch (err) {
      reject(err);
    }
  });
};
