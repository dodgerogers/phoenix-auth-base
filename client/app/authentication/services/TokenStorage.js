import moment from 'moment';


export const STORAGE_KEY = 'token';
const storage = global.localStorage;

export const encode = token => btoa(JSON.stringify(token));
export const generateExpiry = token => moment(token.createdAt).add(token.expiresIn, 'seconds').toDate();

export const decode = token => {
  try {
    return JSON.parse(atob(token));
  } catch (err) {
    return null;
  }
};

export const store = token => {
  return new Promise((resolve, reject) => {
    try {
      storage.setItem(STORAGE_KEY, encode(token));
      resolve();
    } catch (err) {
      reject(err);
    }
  });
}

export const fetch = () => {
  return new Promise((resolve, reject) => {
    const foundToken = storage.getItem(STORAGE_KEY);
    if (foundToken === null) {
      return reject(null);
    }

    const decodedToken = decode(foundToken);
    return resolve(decodedToken);
  });
};

export const remove = () => {
  return new Promise((resolve, reject) => {
    try {
      storage.removeItem(STORAGE_KEY);
      resolve(null);
    } catch (err) {
      reject(err);
    }
  });
};
