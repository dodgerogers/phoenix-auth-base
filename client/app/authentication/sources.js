import HTTP, { API_BASE } from '../lib/utils/HTTP';

export const login = ({ email, password }) => {
  return HTTP.post(`${API_BASE}/api/sessions`, { session: { email, password } });
};

export const register = ({ name, email, password, password_confirmation }) => {
  return HTTP.post(`${API_BASE}/api/registrations`, { registration: { name, email, password, password_confirmation } });
};
