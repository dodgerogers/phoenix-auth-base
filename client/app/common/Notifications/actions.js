import { actionTypes } from './constants';


export const notify = (message, areaID) => ({
  type: actionTypes.NOTIFY,
  id: areaID,
  notification: {
    level: 'success',
    message,
  },
});

export const destroy = (notification, areaID) => ({
  type: actionTypes.DESTROY,
  id: areaID,
  notification,
});
