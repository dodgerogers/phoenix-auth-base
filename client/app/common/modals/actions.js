import { actionTypes } from './constants';

export function showModal(id) {
  return {
    type: actionTypes.SHOW_MODAL,
    data: { id },
  };
}

export function hideModal(id) {
  return {
    type: actionTypes.HIDE_MODAL,
    data: { id },
  };
}
