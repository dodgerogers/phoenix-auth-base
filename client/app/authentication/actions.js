import { stopAsyncValidation, SubmissionError } from 'redux-form/immutable';
import * as AuthenticationSources from './sources';
import { ModalActions, ModalIds } from '../common/modals';
import { actionTypes, formIDs } from './constants';

const authenticateSuccess = (user) => ({
  type: actionTypes.AUTHENTICATE_SUCCESS,
  user,
});

const authenticateFailure = (error) => ({
  type: actionTypes.AUTHENTICATE_FAILURE,
  error,
});

export function login(loginParams) {
  return dispatch => {
    return AuthenticationSources.login(loginParams.toJS())
      .then(response => {
        dispatch(authenticateSuccess(response.data));
        dispatch(ModalActions.hideModal(ModalIds.loginModal));
      })
      .catch(err => {
        dispatch(authenticateFailure(err.response.data.error))
        throw new SubmissionError({ _error: err.response.data.error });
      });
  };
}

const registerSuccess = (user) => ({
  type: actionTypes.REGISTER_SUCCESS,
  user,
});

const registerFailure = (error) => ({
  type: actionTypes.REGISTER_FAILURE,
  error,
});

export function register(registerParams) {
  return dispatch => {
    return AuthenticationSources.register(registerParams.toJS())
      .then(response => {
        dispatch(registerSuccess(response.data))
      })
      .catch(err => {
        dispatch(registerFailure(err.response.data.error))
        throw new SubmissionError(err.response.data.error);
      });
  };
}
