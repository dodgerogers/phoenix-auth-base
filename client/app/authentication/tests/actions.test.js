import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import HTTP from '../../lib/utils/HTTP';
import MockAdapter from 'axios-mock-adapter';
import { fromJS } from 'immutable';
import normalize from 'normalize-object';
import * as AuthenticationActions from '../actions';
import * as AuthenticationSources from '../sources';
import * as TokenStorage from '../lib/TokenStorage';


const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('async AuthenticationActions', () => {
  let mockAxios;
  beforeAll(() => {
    mockAxios = new MockAdapter(HTTP);
  });

  afterEach(() => {
    mockAxios.reset();
  });

  afterAll(() => {
    mockAxios.restore();
  });

  describe('authenticate', () => {
    afterEach(() => {
      TokenStorage.fetch.mockReset();
    });

    it('when TokenStore.fetch is successful', () => {
      const mockToken = { accessToken: "token" };
      TokenStorage.fetch = jest.fn(() => Promise.resolve(mockToken));

      const store = mockStore();

      return store.dispatch(AuthenticationActions.authenticate())
        .then(() => {
          expect(store.getActions()).toMatchSnapshot();
        });
    });

    it('when TokenStore.fetch fails', () => {
      TokenStorage.fetch = jest.fn(() => Promise.reject("Error"));

      const store = mockStore();

      return store.dispatch(AuthenticationActions.authenticate())
        .then((err) => {
          expect(store.getActions()).toMatchSnapshot();
        });
    });
  });

  describe('signOut', () => {
    it('when AuthenticationSources.signOut is successful', () => {
      mockAxios.onDelete('api/oauth/token')
        .reply(204, {});

      const store = mockStore();

      return store.dispatch(AuthenticationActions.signOut())
        .then(() => {
          expect(store.getActions()).toMatchSnapshot();
        });
    });

    it('when AuthenticationSources.signOut fails', () => {
      const error = 'Something went wrong';
      mockAxios.onDelete('api/oauth/token')
        .reply(400, { error });

      const store = mockStore();

      return store.dispatch(AuthenticationActions.signOut())
        .then(() => {
          expect(store.getActions()).toMatchSnapshot();
        });
    });
  });

  describe('register', () => {
    let args;
    beforeEach(() => {
      const password = 'password';
      args = {
        name: 'Bob',
        email: 'email@email.com',
        password,
        passwordConfirmation: password,
      };
    });

    it('when registration is successful', () => {
      const mockResponse = { id: 1, name: args.name, email: args.email };
      mockAxios.onPost('api/registrations', { registration: normalize(args, 'snake') })
        .reply(200, mockResponse);

      const store = mockStore();
      const params = fromJS(args);

      return store.dispatch(AuthenticationActions.register(params))
        .then(() => {
          expect(store.getActions()).toMatchSnapshot();
        });
    });

    it('when registration fails', () => {
      const mockResponse = {
        error: {
          email: 'Has already been taken',
        }
      };
      mockAxios.onPost('api/registrations', { registration: normalize(args, 'snake') })
        .reply(400, mockResponse);

      const store = mockStore();
      const params = fromJS(args);

      return store.dispatch(AuthenticationActions.register(params))
        .catch(() => {
          expect(store.getActions()).toMatchSnapshot();
        });
    });
  });

  describe('confirmation', () => {
    let args;
    beforeEach(() => {
      const email = 'email@email.com';
      const password = 'password';
      const confirmationToken = 'token';
      args = {
        email,
        password,
        confirmationToken: confirmationToken
      };
    });

    it('when confirmation is successful', () => {
      const mockResponse = { id: 1, name: args.name, email: args.email };
      mockAxios.onPut('api/confirmations', { confirmation: normalize(args, 'snake') })
        .reply(200, mockResponse);

      const store = mockStore();
      const params = fromJS(args);

      return store.dispatch(AuthenticationActions.confirm(params))
        .then(() => {
          expect(store.getActions()).toMatchSnapshot();
        });
    });

    it('when confirmation fails', () => {
      const mockResponse = {
        error: "Invalid credentials",
      };
      mockAxios.onPut('api/confirmations', { confirmation: normalize(args, 'snake') })
        .reply(400, mockResponse);

      const store = mockStore();
      const params = fromJS(args);

      return store.dispatch(AuthenticationActions.confirm(params))
        .catch(() => {
          expect(store.getActions()).toMatchSnapshot();
        });
    });
  });

  describe('login', () => {
    let email, password;
    beforeAll(() => {
      email = 'email@email.com';
      password = 'password';
    });

    it('when login is successful', () => {
      const mockTokenResponse = {
        message: "Login successful",
        accessToken: {
          accessToken: "code",
        }
      };

      mockAxios.onPost('api/oauth/token', { grant_type: 'password', username: email, password })
        .reply(200, mockTokenResponse);

      const mockUserResponse = {
        id: 1,
        name: 'name',
      }

      const store = mockStore();
      const params = fromJS({ email, password });

      return store.dispatch(AuthenticationActions.login(params))
        .then(() => {
          expect(store.getActions()).toMatchSnapshot();
        });
    });

    it('when login fails', () => {
      const mockResponse = {
        error: 'Invalid credentials',
      };

      mockAxios.onPost('api/oauth/token', { grant_type: 'password', username: email, password })
        .reply(401, mockResponse);

      const store = mockStore();
      const params = fromJS({ email, password });

      return store.dispatch(AuthenticationActions.login(params))
        .catch(() => {
          expect(store.getActions()).toMatchSnapshot();
        });
    });
  });

  describe('resendConfirmation', () => {
    let args;
    beforeEach(() => {
      const email = 'email@email.com';
      args = {
        email,
      };
    });

    it('when resendConfirmation is successful', () => {
      const mockResponse = { message: "Success" };
      mockAxios.onPost('api/confirmations', { confirmation: args })
        .reply(200, mockResponse);

      const store = mockStore();
      const params = fromJS(args);

      return store.dispatch(AuthenticationActions.resendConfirmation(params))
        .then(() => {
          expect(store.getActions()).toMatchSnapshot();
        });
    });

    it('when confirmation fails', () => {
      const mockResponse = {
        error: "Invalid credentials",
      };
      mockAxios.onPost('api/confirmations', { confirmation: args })
        .reply(400, mockResponse);

      const store = mockStore();
      const params = fromJS(args);

      return store.dispatch(AuthenticationActions.resendConfirmation(params))
        .catch(() => {
          expect(store.getActions()).toMatchSnapshot();
        });
    });
  });
});
