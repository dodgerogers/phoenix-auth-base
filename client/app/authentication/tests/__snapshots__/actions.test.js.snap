// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`async AuthenticationActions login when login fails 1`] = `
Array [
  Object {
    "error": "Invalid credentials",
    "type": "AUTHENTICATE_FAILURE",
  },
]
`;

exports[`async AuthenticationActions login when login is successful 1`] = `
Array [
  Object {
    "type": "AUTHENTICATE_SUCCESS",
    "user": Object {
      "id": 1,
      "name": "name",
    },
  },
  Object {
    "data": Object {
      "id": "loginModal",
    },
    "type": "HIDE_MODAL",
  },
]
`;

exports[`async AuthenticationActions register when registration fails 1`] = `
Array [
  Object {
    "error": Object {
      "email": "Has already been taken",
    },
    "type": "REGISTER_FAILURE",
  },
]
`;

exports[`async AuthenticationActions register when registration is successful 1`] = `
Array [
  Object {
    "type": "REGISTER_SUCCESS",
    "user": Object {
      "email": "email@email.com",
      "id": 1,
      "name": "Bob",
    },
  },
]
`;
